TPC/TPCbase/AliTPC3DCylindricalInterpolatorIrregular.cxx

```diff
@@ -0,0 +1,1167 @@
+  Int_t new_rIndex = 0;
+  Int_t new_zIndex = 0;
+  Int_t new_phiIndex = 0;
+
+  for (Int_t iPhi = startPhi; iPhi < startPhi + phiStep; iPhi++) {
+    indexPhi = iPhi % fNPhi;
+    for (Int_t index_r = startR; index_r < startR + rStep; index_r++) {
+      for (Int_t index_z = startZ; index_z < startZ + zStep; index_z++) {
+        // check for the closest poInt_t
+        index = indexPhi * (fNZ * fNR) + index_r * fNZ + index_z;
+
+        r0 = fRList[index];
+        z0 = fZList[index];
+        phi0 = fPhiList[index];
+
+        d = Distance(r0, phi0, z0, r, phi, z);
+        if (d < shortest_d) {
+          shortest_d = d;
+          new_rIndex = index_r;
+          new_phiIndex = indexPhi;
+          new_zIndex = index_z;
+        }
+      }
+    }
+  }
+
+  phiStep = 3;
+  rStep = 3;
+  startPhi = new_phiIndex - phiStep / 2;
+  startR = new_rIndex - rStep / 2;
+  startZ = new_zIndex - zStep / 2;
+
+  if (startPhi < 0) startPhi = fNPhi + startPhi;
+  if (startR < 0) startR = 0;
+  if (startR + rStep >= fNR) startR = fNR - rStep;
+
+  if (startZ < fMinZIndex) startZ = fMinZIndex;
+  if (startZ + zStep >= fNZ) startZ = fNZ - zStep;
+
+  for (Int_t iPhi = startPhi; iPhi < startPhi + phiStep; iPhi++) {
+    indexPhi = iPhi % fNPhi;
+    for (Int_t index_r = startR; index_r < startR + rStep; index_r++) {
+      for (Int_t index_z = startZ; index_z < startZ + zStep; index_z++) {
+        // check for the closest poInt_t
+        index = indexPhi * (fNZ * fNR) + index_r * fNZ + index_z;
+
+        r0 = fRList[indexPhi * (fNR * fNZ) + index_r * fNZ + index_z];
+        z0 = fZList[indexPhi * (fNR * fNZ) + index_r * fNZ + index_z];
+        phi0 = fPhiList[indexPhi * (fNR * fNZ) + index_r * fNZ + index_z];
+        d = Distance(r0, phi0, z0, r, phi, z);
+        if (d < MIN_DIST) {
+          return fValue[index];
+        }
+        d = 1.0 / d;
+        sum_w += (fValue[index] * d * d * d * d);
+        sum_d += d * d * d * d;
+      }
+    }
+  }
+  return (sum_w / sum_d);
+}
+
+/// distance in Cyl coordinate
+///
+/// \param r0
+/// \param phi0
+/// \param z0
+/// \param r
+/// \param phi
+/// \param z
+/// \return
+Double_t
+AliTPC3DCylindricalInterpolatorIrregular::Distance(Double_t r0, Double_t phi0, Double_t z0, Double_t r, Double_t phi,
+                                                   Double_t z) {
+  Double_t x0 = r0 * TMath::Cos(phi0);
+  Double_t y0 = r0 * TMath::Sin(phi0);
+  Double_t x = r * TMath::Cos(phi);
+  Double_t y = r * TMath::Sin(phi);
+
+  if (phi < 0) phi = TMath::TwoPi() + phi;
+  if (phi > TMath::TwoPi()) phi = phi - TMath::TwoPi();
+
+  if (phi0 < 0) phi0 = TMath::TwoPi() + phi0;
+  if (phi0 > TMath::TwoPi()) phi0 = phi0 - TMath::TwoPi();
+
+  Double_t dPhi = phi - phi0;
+  if (dPhi > TMath::Pi())
+    dPhi = TMath::TwoPi() - dPhi;
+  if (dPhi < -TMath::Pi())
+    dPhi = TMath::TwoPi() + dPhi;
+
+  Double_t ret = (r - r0) * (r - r0) + (dPhi * dPhi) * ((r + r0) / 2.0) + (z - z0) * (z - z0);
```

Comment:
I don't understand this formula, but it is certainly wrong, since the 2 term had dimension of L instead of L^2. The standard formula is r^2+r0^2 - 2 r0*r*cos(dPhi) + (z-z0)^2

Reply:
thanks for the correction, yes I should put distance formula for cylindrical coordinate. as you note.